### PR TITLE
Add user_path and permissions_boundary_arn arguments to aws_secret_backend_role

### LINKED ISSUE #716: vault_aws_secret_backend_role: currently doesn't support specifying permissions_boundary_arn
Hi. Vault 1.3 added support for specifying a permissions boundary when creating AWS roles (https://github.com/hashicorp/vault/pull/6786), but it looks as though Terraform doesn't support the permissions_boundary_arn parameter yet. This role definition:

    resource "vault_aws_secret_backend_role" "test-ec2" {
      backend = "${vault_aws_secret_backend.test.path}"
      name = "test-ec2"
      credential_type = "iam_user"
      permissions_boundary_arn = "arn:aws:iam::0123456789:policy/vault-permissions-boundary"

      policy_document = <<EOT
      {
        "Version": "2012-10-17",
        "Statement": [
          {
            "Effect": "Allow",
            "Action": "ec2:*",
            "Resource": "*"
          }
        ]
      }
      EOT
    }

Throws the following error on apply:

    Error: Unsupported argument on roles-aws.tf line 5, in resource "vault_aws_secret_backend_role" "test-ec2":   
    5:   permissions_boundary_arn = "arn:aws:iam::0123456789:policy/vault-permissions-boundary" 
    An argument named "permissions_boundary_arn" is not expected here.`

Any chance this could be added? Thanks!

### PR DESCRIPTION
**Pull request by bakeemawaytoys**
I modified the aws_secret_backend_role resource to accept the user_path and permissions_boundary_arn arguments.  I updated the resource's tests to account for the new arguments and consolidated some duplicate code in the tests into functions.  I also removed the tests' dependency on AWS credentials because Vault doesn't interact with AWS during CRUD operations on the backend role.  It will now be easier to run the tests on the aws_secret_backend_role resource.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Closes #716, #769 

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
Add `user_path` and `permissions_boundary_arn` fields to `aws_secret_backend_role`.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSecretBackendRole'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccAWSSecretBackendRole -timeout 120m
?       github.com/terraform-providers/terraform-provider-vault [no test files]
?       github.com/terraform-providers/terraform-provider-vault/cmd/coverage    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/util    0.023s [no tests to run]
=== RUN   TestAccAWSSecretBackendRole_basic
--- PASS: TestAccAWSSecretBackendRole_basic (0.27s)
=== RUN   TestAccAWSSecretBackendRole_import
--- PASS: TestAccAWSSecretBackendRole_import (0.23s)
=== RUN   TestAccAWSSecretBackendRole_nested
--- PASS: TestAccAWSSecretBackendRole_nested (0.25s)
PASS
ok      github.com/terraform-providers/terraform-provider-vault/vault   0.772s
...
```

### CODE REVIEW

**Comment by gfalqui** [comment]
Please merge this! I'd love to have this too!

**Comment by gfalqui** [comment]
```hcl
resource "vault_aws_secret_backend_role" "readonly" {
  backend         = vault_aws_secret_backend.aws.path
  name            = "readonly"
  credential_type = "iam_user"

  policy_arns = [
    "arn:aws:iam::aws:policy/ReadOnlyAccess"
  ]

  provisioner "local-exec" {
    command = "vault write -address=${var.vault_endpoint} -namespace=${var.vault_namespace} ${vault_aws_secret_backend.aws.path}/roles/readonly user_path=\"/vault-users/${local.vault_namespace_with_dashes}/\" permissions_boundary_arn=\"arn:aws:iam::${data.aws_caller_identity.current.account_id}:policy/policy-boundaries-name\""
  }
}
```

Using a provisioner can be a workaround! 

Just make sure the environment contains Vault in the $PATH that has got a version that supports the new parameters.

**Comment by cyriptix** [comment]
> Using a provisioner can be a workaround!

Worked a treat, thank you! 👍

**Comment by billyaustin84** [comment]
Is this likely to make it to the top of the pile for merging?

**Comment by bakeemawaytoys** [comment]
I think I was supposed to sign the CLA when I first submitted this PR but failed to do so.  A helpful bot has alerted me to the situation.  I've signed it and rebased to address the conflicts.  Sorry about the delay on my end.

**Comment by billyaustin84** [comment]
Could I ask again when this is likely to be reviewed and merged?

**Comment by benashz** [comment]
Hi @bakeemawaytoys , 

We'd like to take another look at this PR. If you still have interest, would you mind rebasing it off of main. If not we would be happy to take it over.

Thanks,

Ben

**Comment by bakeemawaytoys** [comment]
Sure thing. I should have some time to rebase it this weekend.

**Comment by bakeemawaytoys** [comment]
Some of my changes to the test file made for a tricky rebase, but I think it is all set.

**Comment by benashz** [comment]
@bakeemawaytoys thanks!

**Comment by benashz** [comment]
Hi @bakeemawaytoys, it looks like this PR has broken some of the acceptance tests (see below for an example). Please let us know if you an assist here.

Thanks,

Ben

```
=== RUN   TestAccAWSSecretBackendRole_nested
    resource_aws_secret_backend_role_test.go:100: Step 1/2 error: Error running pre-apply refresh: exit status 1
        
        Error: Missing item separator
        
          on terraform_plugin_test.tf line 19, in resource "vault_aws_secret_backend_role" "test_policy_arns":
          19:   policy_arns = ["{"Version": "2012-10-17","Statement": [{"Effect": "Allow","Action": "iam:*","Resource": "*"}]}"]
        
        Expected a comma to mark the beginning of the next item.
    testing_new.go:63: Error retrieving state, there may be dangling resources: exit status 1
        
        Error: Missing item separator
        
          on terraform_plugin_test.tf line 19, in resource "vault_aws_secret_backend_role" "test_policy_arns":
          19:   policy_arns = ["{"Version": "2012-10-17","Statement": [{"Effect": "Allow","Action": "iam:*","Resource": "*"}]}"]
        
        Expected a comma to mark the beginning of the next item.
        
--- FAIL: TestAccAWSSecretBackendRole_nested (0.66s)
```

**Comment by bakeemawaytoys** [comment]
@benashz I fixed the failing tests and a few others that revealed themselves after I fixed the initial issue.  Sorry for not catching that myself.

File: vault/resource_aws_secret_backend_role.go
Comment (benashz) [inline vault/resource_aws_secret_backend_role.go:133]: It might be a good idea to log a warning when the boundary ARN is configured and the `credential_type` is not `iam_user`.

File: vault/resource_aws_secret_backend_role.go
Comment (benashz) [inline vault/resource_aws_secret_backend_role.go:148]: It might be a good idea to log a warning when the user path is configured and the `credential_type` is not `iam_user`.

**Review by benashz — COMMENTED** [review]
Looks good. I only have a few comments/suggestions that need to be addressed before approval.

**Review by benashz — APPROVED** [review]
LGTM!

Thanks again for your contribution to HashiCorp!

**Pull request merged**
